### PR TITLE
support ECMA 5 non-conformant behaviour of Microsoft edge

### PR DESCRIPTION
**Pull request by freddiecoleman**
Microsoft Edge does not conform to the ECMA 5 spec (http://www.ecma-international.org/ecma-262/5.1/#sec-15.2.4.2).

toString of null returns [Object Window].

This patch fixes the issue and adds tests for these edge cases.

### CODE REVIEW

**Comment by claylaut** [comment]
+1 - this will also fix IOS 9 and below.

**Comment by flesler** [comment]
Thanks @freddiecoleman, is there a way we can add a test for this?

**Comment by flesler** [comment]
Hi @freddiecoleman, my bad on the test comment, here are a few things that it'd be great if you can tweak. I prefer if you squash all in a single commit (althought I think I can do that now when merging the PR now)
1. Remove needless indents on test.js
2. Fix wrong indentation on test.js, several blocks have different indentation than the rest
3. If possible, please bump the package.json version to `2.1.0` so there's no need for an additional commit from me just for that
4. The `if` is missing parens, it should be `!key && (type === 'domwindow' || type === 'window')` but it could reduced to `!key && type.slice(-6) === 'window'`, I prefer the latter but either would be fine.
5. Missing a semicolon [here](https://github.com/freddiecoleman/hashmap/blob/3b51dc8dbdb3dd209b48cc57d2af41797e4a7367/test/test.js#L39).
6. Needless parameter [here](https://github.com/freddiecoleman/hashmap/blob/3b51dc8dbdb3dd209b48cc57d2af41797e4a7367/test/test.js#L37)

Rest is fine, I like the clever way you came with to test this. About the indentation comments, I should add an `.editorconfig` file to make it more explicit but for now, following the implicit standard is the desirable approach.

File: hashmap.js [hunk 1]
```diff
@@ -85,8 +85,8 @@
 		type:function(key) {
 			var str = Object.prototype.toString.call(key);
 			var type = str.slice(8, -1).toLowerCase();
-			// Some browsers yield DOMWindow for null and undefined, works fine on Node
-			if (type === 'domwindow' && !key) {
+			// Some browsers yield DOMWindow or Window for null and undefined, works fine on Node
+			if (!key && type === 'domwindow' || type === 'window') {
```
Comment (flesler) [inline hashmap.js]: The if is missing parens, it should be !key && (type === 'domwindow' || type === 'window') but it could reduced to !key && type.slice(-6) === 'window', I prefer the latter but either would be fine.

File: test/test.js [hunk 2]
```diff
@@ -29,6 +29,24 @@ describe('hashmap', function() {
...
 			check(hashmap, 'object');
 		});
+
+        it('should cast type for browsers that don\'t conform to ECMA 5 spec', function() {
+            var originalToString = Object.prototype.toString;
+            var type = 'DOMWindow';
+
+            Object.prototype.toString = function(a) {
+                return '[object ' + type + ']';
+            }
```
Comment (flesler) [inline test/test.js]: Missing semicolon

File: test/test.js [hunk 3]
```diff
@@ -29,6 +29,24 @@ describe('hashmap', function() {
 			check(HashMap, 'function');
 			check(hashmap, 'object');
 		});
+
+        it('should cast type for browsers that don\'t conform to ECMA 5 spec', function() {
+            var originalToString = Object.prototype.toString;
+            var type = 'DOMWindow';
+
+            Object.prototype.toString = function(a) {
```
Comment (flesler) [inline test/test.js]: Needless parameter

File: test/test.js [hunk 4]
```diff
@@ -29,6 +29,24 @@ describe('hashmap', function() {
 			check(HashMap, 'function');
 			check(hashmap, 'object');
 		});
+
+        it('should cast type for browsers that don\'t conform to ECMA 5 spec', function() {
```
Comment (flesler) [inline test/test.js]: This block is incorrectly indented, please indent as the rest

File: test/test.js [hunk 5]
```diff
@@ -132,7 +150,7 @@ describe('hashmap', function() {
 				hashmap.set(key, value);
 				expect(hashmap.get(key)).to.equal(value);
 			}
-			
+
```
Comment (flesler) [inline test/test.js]: No need to change this line

File: test/test.js [hunk 6]
```diff
@@ -399,7 +417,7 @@ describe('hashmap', function() {
 		it('should clone a hashmap when one argument', function() {
 			hashmap.set('key', 'value');
 			hashmap.set('key2', 'value2');
-			
+
```
Comment (flesler) [inline test/test.js]: No need to change this line

File: test/test.js [hunk 7]
```diff
@@ -419,7 +437,7 @@ describe('hashmap', function() {
 
 	afterEach(function() {
 	});
-	
+
```
Comment (flesler) [inline test/test.js]: No need to change this line

**Comment by freddiecoleman** [comment]
I have implemented the feedback from the review, including bumping the version, and combined it into a single commit.
